### PR TITLE
Please, merge my forked repo. Many improvements.

### LINKED ISSUE #24: Twitter: timeline gets reposted
Sometimes the whole timeline gets reposted on the channel. Even tweets that were already published.

We should investigate the asynchronous classes: 
- http://twitter4j.org/en/javadoc/twitter4j/AsyncTwitter.html

Or the streaming API:
- http://twitter4j.org/javadoc/twitter4j/TwitterStream.html

to fix this problem and avoid reading the latest 20 tweets from the timeline and calculating the new ones every time.

### PR DESCRIPTION
**Pull request by gpoppino**
I have been working on my forked repo for some time and I'd like to merge the changes into the master branch of amnotbot:amnotbot.

There are many improvements and new commands (spotify, new weather command, SSL support, etc). This closes #24 and #6 too.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+RUN chown -R amnotbot:amnotbot /home/amnotbot
+
+USER amnotbot
+ENV HOME /home/amnotbot
+
+WORKDIR /home/amnotbot/app
+
+RUN mvn -e install -DskipTests
+
+CMD ["java","-jar","target/amnotbot-core-0.0.1-SNAPSHOT.jar"]
```
Comment (jmitchener) [inline Dockerfile]: Can we do this without hardcoding version?

File: src/main/java/com/github/amnotbot/BotCommandInterpreter.java [hunk 2]
```diff
@@ -29,39 +29,32 @@
...
-    public BotCommandInterpreter(BotSpamDetector spamDetector)
+    private HashMap<BotCommandEvent, List<BotCommand>> cmdListeners;
+
+    public BotCommandInterpreter()
     {    
-        this.spamDetector = spamDetector;
-        this.cmdTrigger =
-                BotConfiguration.getConfig().getString("command_trigger", ".");
+        this.spamDetector = new BotSpamDetector();
+        this.helpTrigger =
```
Comment (jmitchener) [inline src/main/java/com/github/amnotbot/BotCommandInterpreter.java]: Why are these on multiple lines?

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+language: java
+jdk:
+    - oraclejdk8
```
Comment (gpoppino) [inline .travis.yml]: For the time being, Travis CI does not offer OpenJDK 8, only OpenJDK 7. If we want to use Java 8, we have to go with OracleJDK 8 for Travis.

I run the bot with OpenJDK 8. But I understand your suggestion, we should run the bot and test it with the same version or in multiple versions in any case.

Reference: https://docs.travis-ci.com/user/languages/java/

File: pom.xml [hunk 4]
```diff
@@ -3,64 +3,90 @@
   <groupId>com.github.amnotbot</groupId>
   <artifactId>amnotbot-core</artifactId>
   <version>0.0.1-SNAPSHOT</version>
+
+  <scm>
```
Comment (gpoppino) [inline pom.xml]: Ok. This is needed by the buildnumber-maven-plugin. 
I will update it to this github.

File: Dockerfile (hunk 1 above)
Comment (gpoppino) [inline Dockerfile]: I tried `ENV` and `ARG` but could not make it work. `ENV` is only visible inside the container, I understand and not during the building process. And `ARG` you need to pass the value when building the image.

Tried something like:

`ENV AMNOTBOT_VERSION 0.0.1-SNAPSHOT`
`CMD ["java","-jar","target/amnotbot-core-$AMNOTBOT_VERSION"]`

But without success (in docker for openSUSE 42.2).

File: src/main/java/com/github/amnotbot/BotCommandInterpreter.java (hunk 2 above)
Comment (gpoppino) [inline src/main/java/com/github/amnotbot/BotCommandInterpreter.java]: Well, I tried not to go beyond famous column number 80. If that is what you are referring to.

File: pom.xml [hunk 5]
```diff
@@ -83,35 +109,40 @@
...
+      <groupId>commons-configuration</groupId>
+      <artifactId>commons-configuration</artifactId>
+      <version>1.10</version>
     </dependency>
     <dependency>
         <groupId>org.twitter4j</groupId>
-        <artifactId>twitter4j-core</artifactId>
-        <version>3.0.3</version>
+        <artifactId>twitter4j-stream</artifactId>
+        <version>[4.0,)</version>
```
Comment (gpoppino) [inline pom.xml]: I think it does not hurt to use 4.0.4 or 4.0.5. They are just minor updates. However, to be sure it works, we can stay on a fixed version.

Regarding the pulling of version numbers out to properties, we can discuss this on IRC so you can clarify some doubts I have.

Comment (jmitchener) [inline pom.xml]: It doesn't need to be hardcoded. I meant this has no upper bound. It should be [4.0,5) so it doesn't update beyond 4.x. Possibly even [4.0,4.1). We should define an upper bound here.

**Comment by jmitchener** [comment]
Still has lots of unaddressed issues but I think we can just merge after 3 years ... we can make issues for things

**Pull request merged**